### DISCUSSION TITLE
Centralized version description of modules

### DISCUSSION THREAD

**Discussion by MageParts** [PrestaShop/PrestaShop#40249] [Ideas]
#### Is your feature request related to a problem?

Version currently needs to be defined by both composer.json and module base file.

People will confuse these values and most likely forget to update the version number within the base file. At the very least it's an unnecessary annoyance which is very easily solved.

#### Describe the solution you'd like

PrestaShop should omit version assignment from the constructor in the base file and resolve the value from the module composer.json file instead.

Since this would require all modules to be converted to proper composer packages it's of course not completely viable, at least for now, but you could include a class like the one I've described below in a **ps_composer** module or likewise to simplify this for new development:

```
/**
 * Read data from composer.json
 */
class Composer {
    /**
     * @var string
     */
    private $file;

    /**
     * @param string $dir
     */
    public function __construct(string $dir) {
        $this->file = "$dir/composer.json";
    }

    /**
     * @return string
     * @throws ComposerException
     */
    public function getComposerConfigFile(): string {
        if (!file_exists($this->file)) {
            throw new ComposerException('Missing module composer.json file.');
        }

        return $this->file;
    }

    /**
     * @return array
     * @throws JsonException
     * @throws ComposerException
     */
    public function getComposerConfig(): array {
        $result = [];
        $content = file_get_contents($this->getComposerConfigFile());

        if (is_string($content) && $content !== '') {
            $result = json_decode(
                $content,
                true,
                512,
                JSON_THROW_ON_ERROR
            );

            if (!is_array($result)) {
                throw new ComposerException('Failed to parse composer.json');
            }
        }

        return $result;
    }

    /**
     * @return string
     * @throws JsonException
     * @throws ComposerException
     */
    public function getVersion(): string {
        $config = $this->getComposerConfig();

        if (!isset($config['version'])) {
            throw new ComposerException('Please specify version in module composer.json');
        }

        if (!is_string($config['version'])) {
            throw new ComposerException('Invalid version value in module composer.json');
        }

        return $config['version'];
    }

    /**
     * @param int $index
     * @return string
     * @throws ComposerException
     * @throws JsonException
     */
    public function getAuthor(int $index = 0): string {
        $config = $this->getComposerConfig();

        if (!isset($config['authors'])) {
            throw new ComposerException('Please specify authors in module composer.json');
        }

        if (!is_array($config['authors'])) {
            throw new ComposerException('Invalid authors value in module composer.json, must be array.');
        }

        if (!isset($config['authors'][$index]['name'])) {
            throw new ComposerException("Missing name key on author index $index composer.json");
        }

        return $config['authors'][$index]['name'];
    }

    /**
     * @return string
     * @throws JsonException
     * @throws ComposerException
     */
    public function getDescription(): string {
        $config = $this->getComposerConfig();

        if (!isset($config['description'])) {
            throw new ComposerException('Please specify description in module composer.json');
        }

        if (!is_string($config['description'])) {
            throw new ComposerException('Invalid description value in module composer.json');
        }

        return $config['description'];
    }
}
```

NOTE: **ComposerException** is simply a class extending **\Exception** in this case.
NOTE: The class above has been tested, I use it myself in a couple of modules to avoid having to specify the module name/version/author in multiple locations.
NOTE: An idea would be to have PrestaShop:s Module base class naturally fallback to use the values from the modules composer.json file if it exists and the values have not been assigned manually in the constructor (ie. I didn't set $this->version in my Module constructor, so resolve it from my composer.json).

**Comment by khouloudbelguith** [comment]
Hi @MageParts,

Thank you for your suggestion. The Product Team will take it into consideration for future developments. 
Please be aware that there is no guarantee that this feature will be developed anytime soon. If this is important to you, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you create it. 

If you are willing to implement this feature into PrestaShop as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/), the teams will be happy to help you by ensuring that it is appropriate and coherent with the rest of the project. If that is the case, please let us know! 

Thank you!

**Comment by MageParts** [comment]
Hi @khouloudbelguith!

As stated in my suggestion I've already implemented this class for my own modules to ensure version information is accurate in them. There is no more development required than to add the class I supplied above to a module. I understand you cannot guarantee it will be added of course, though I would recommend you internally discuss the necessity of duplicate version numbers between the composer configuration file, and the module base file.

Thank you for taking my proposal under consideration, I would have PR:ed if I knew where to put it which unfortunately I don't.

Again thank you and have a good weekend :-)
